modules/python/python-types.c

```diff
@@ -0,0 +1,446 @@
+{
+  PyObject *obj = NULL;
+  const gchar *charset;
+
+  /* NOTE: g_get_charset() returns if the current character set is utf8 */
+  if (g_get_charset(&charset))
+    {
+      if (len < 0)
+        obj = PyUnicode_FromString(value);
+      else
+        obj = PyUnicode_FromStringAndSize(value, len);
+    }
+  else
+    {
+      GError *error = NULL;
+      gsize bytes_read, bytes_written;
+      gchar *utf8_string;
+
+      utf8_string = g_locale_to_utf8(value, len, &bytes_read, &bytes_written, &error);
+      if (utf8_string)
+        {
+          obj = PyUnicode_FromStringAndSize(utf8_string, bytes_written);
+          g_free(utf8_string);
+        }
+      else
+        {
+          g_error_free(error);
+          if (len >= 0)
+            obj = PyBytes_FromStringAndSize(value, len);
+          else
+            obj = PyBytes_FromString(value);
+        }
+    }
+
+  if (!obj)
+    _log_py_object_construction_error();
+
+  return obj;
+}
+
+PyObject *
+py_long_from_long(gint64 l)
+{
+  PyObject *obj = PyLong_FromLong(l);
+  if (!obj)
+    _log_py_object_construction_error();
+
+  return obj;
+}
+
+PyObject *
+py_double_from_double(gdouble d)
+{
+  PyObject *obj = PyFloat_FromDouble(d);
+  if (!obj)
+    _log_py_object_construction_error();
+
+  return obj;
+}
+
+PyObject *
+py_boolean_from_boolean(gboolean b)
+{
+  PyObject *obj = PyBool_FromLong(b);
+  if (!obj)
+    _log_py_object_construction_error();
+
+  return obj;
+}
+
+PyObject *
+py_list_from_list(const gchar *list, gssize list_len)
+{
+  PyObject *obj = PyList_New(0);
+  if (!obj)
+    {
+      _log_py_object_construction_error();
+      return NULL;
+    }
+
+  ListScanner scanner;
+  list_scanner_init(&scanner);
+  list_scanner_input_string(&scanner, list, list_len);
+
+  while (list_scanner_scan_next(&scanner))
+    {
+      PyObject *element = py_bytes_from_string(list_scanner_get_current_value(&scanner),
+                                               list_scanner_get_current_value_len(&scanner));
+      if (!element || PyList_Append(obj, element) != 0)
+        {
+          list_scanner_deinit(&scanner);
+          _log_py_object_construction_error();
+          Py_DECREF(obj);
+          return NULL;
+        }
+    }
+
+  list_scanner_deinit(&scanner);
+
+  return obj;
+}
+
+PyObject *
+py_obj_from_log_msg_value(const gchar *value, gssize value_len, LogMessageValueType type)
+{
+  switch (type)
+    {
+    case LM_VT_INT32:
+    case LM_VT_INT64:
+    {
+      gint64 l;
+      if (type_cast_to_int64(value, &l, NULL))
+        return py_long_from_long(l);
+      return NULL;
+    }
+
+    case LM_VT_DOUBLE:
+    {
+      gdouble d;
+      if (type_cast_to_double(value, &d, NULL))
+        return py_double_from_double(d);
+      return NULL;
+    }
+
+    case LM_VT_BOOLEAN:
+    {
+      gboolean b;
+      if (type_cast_to_boolean(value, &b, NULL))
+        return py_boolean_from_boolean(b);
+      return NULL;
+    }
+
+    case LM_VT_NULL:
+      return Py_None;
```

Comment:
This should be `Py_RETURN_NONE` (it returns with `Py_None` but does decrease the ref)

Reply:
Thanks, fixed.